pkg/controller/studyjob/studyjob_controller.go

```diff
@@ -240,44 +240,66 @@ func (r *ReconcileStudyJobController) Reconcile(request reconcile.Request) (reco
 		return r.updateFinalizers(instance, finalizers)
 	}
 
+	var newTrials []katibv1alpha1.TrialSet
 	switch instance.Status.Condition {
 	case katibv1alpha1.ConditionCompleted,
 		katibv1alpha1.ConditionFailed,
 		katibv1alpha1.ConditionRunning:
-		update, err = r.checkStatus(instance, request.Namespace)
+		update, newTrials, err = r.checkStatus(instance, request.Namespace)
+		if err != nil {
+			log.Printf("Fail to check status %v", err)
+		}
 	default:
 		now := metav1.Now()
 		instance.Status.StartTime = &now
 		err = initializeStudy(instance)
 		if err != nil {
-			r.Update(context.TODO(), instance)
 			log.Printf("Fail to initialize %v", err)
-			return reconcile.Result{}, err
 		}
 		update = true
 	}
 	now := metav1.Now()
 	instance.Status.LastReconcileTime = &now
-	if err != nil {
-		r.Update(context.TODO(), instance)
-		log.Printf("Fail to check status %v", err)
-		return reconcile.Result{}, err
-	}
-	if update {
-		err = r.Update(context.TODO(), instance)
-		if err != nil {
+	ret := reconcile.Result{}
+	var updateErr error
+	if update || err != nil {
+		updateErr := r.Update(context.TODO(), instance)
```

Comment:
What is the root cause of the race condition?

Reply:
I heard that some jobs can finish meanwhile and the StudyJob get updated.
I've never seen the actual failure though.

Reply:
Like this https://github.com/kubeflow/katib/issues/166#issue-359780752
`2018/09/13 06:44:09 Fail to Update StudyJob c1e236530496ef99 : Operation cannot be fulfilled on studyjobs.kubeflow.org "grid-example": the object has been modified; please apply your changes to the latest version and try again`

When any updates worker or metrics-collector happen during reconcile is running, the update for studyjob will be failed.